safenode/src/domain/storage/disk_backed_record_store.rs

```diff
@@ -0,0 +1,483 @@
+        let mut hex_string = String::with_capacity(key_bytes.len() * 2);
+        for byte in key_bytes {
+            hex_string.push_str(&format!("{:02x}", byte));
+        }
+        hex_string
+    }
+}
+
+impl RecordStore for DiskBackedRecordStore {
+    type RecordsIter<'a> = vec::IntoIter<Cow<'a, Record>>;
+    type ProvidedIter<'a> = vec::IntoIter<Cow<'a, ProviderRecord>>;
+
+    fn get(&self, k: &Key) -> Option<Cow<'_, Record>> {
+        // When a client calls GET, the request is forwarded to the nodes until one node returns
+        // with the record. Thus a node can be bombarded with GET reqs for random keys. These can be safely
+        // ignored if we don't have the record locally.
+        trace!("GET request for Record key: {k:?}");
+        if !self.disk_backed_records.contains(k) {
+            trace!("Record not found locally");
+            return None;
+        }
+
+        let filename = Self::key_to_hex(k);
+        let file_path = self.config.storage_dir.join(&filename);
+
+        if !file_path.exists() {
+            // we went out of sync with the filesystem
+            error!(
+                "Data not found for the provided key, filename: {filename} should exist locally"
+            );
+            return None;
+        }
+
+        match fs::read(file_path) {
+            Ok(contents) => {
+                trace!("Retrieved record from disk! filename: {filename}");
+                let record = Record {
+                    key: k.clone(),
+                    value: contents,
+                    publisher: None,
+                    expires: None,
+                };
+                Some(Cow::Owned(record))
+            }
+            Err(err) => {
+                error!("Error while reading file. filename: {filename}, error: {err:?}");
+                None
+            }
+        }
+    }
+
+    fn put(&mut self, r: Record) -> Result<()> {
+        trace!("PUT request for Record key: {:?}", r.key);
+        if r.value.len() >= self.config.max_value_bytes {
+            warn!(
+                "Record not stored. Value too large: {} bytes",
+                r.value.len()
+            );
+            return Err(Error::ValueTooLarge);
+        }
+
+        // todo: the key is not tied to the value it contains, hence the value can be overwritten
+        // (incase of dbc double spends etc), hence need to deal with those.
+        // Maybe implement a RecordHeader to store the type of data we're storing?
+        if self.disk_backed_records.contains(&r.key) {
+            debug!(
+                "Record with key {:?} already exists, not overwriting.",
+                r.key
+            );
+            return Ok(());
+        }
+
+        let num_records = self.disk_backed_records.len();
+        if num_records >= self.config.max_records {
+            warn!("Record not stored. Maximum number of records reached. Current num_records: {num_records}");
+            return Err(Error::MaxRecords);
```

Comment:
`Error::RecordsTooMany` to be consistent with `Error::ValueTooLarge`

Reply:
Ah, I thought these are errors defined by us.
If they are the `libp2p` ones, then just stay with them.

Reply:
I don't think we can/should here. It should be converted at the upper layers. But to remain within the `RecordStore` trait we'll have to leave this I think